File: ToWer.DbWrapper/MsSql/MsSqlWrapper.cs
using System;
using System.Collections.Generic;
using System.Data.SqlClient;
using System.Dynamic;
using System.Linq;
using System.Text;
using ToWer.DbWrapper.Converter;
using ToWer.DbWrapper.Exceptions;

namespace ToWer.DbWrapper.MsSql
{
    public class MsSqlWrapper : IDbWrapper
    {
        #region Properties

        private string _connectionString;
        private List<SqlParameter> _standardCommandParameters;

        #endregion Properties

        #region Public methods

        public void SetConnectionString(string connectionString)
        {
            _connectionString = connectionString;
        }

        public void SetStandardCommandParameters(List<SqlParameter> parameters)
        {
            if (_standardCommandParameters != null)
            {
                _standardCommandParameters.Clear();
            }
            _standardCommandParameters = parameters;
        }

        public T ReadSingle<T>(string procedureName) where T : class
        {
            if (string.IsNullOrEmpty(_connectionString)) { throw new ArgumentNullException("_connectionString"); }
            return ReadSingle<T>(_connectionString, procedureName, (List<SqlParameter>)null);
        }

        public T ReadSingle<T>(string procedureName, Dictionary<string, object> parameters) where T : class
        {
            if (string.IsNullOrEmpty(_connectionString)) { throw new ArgumentNullException("_connectionString"); }
            return ReadSingle<T>(_connectionString, procedureName, parameters);
        }

        public T ReadSingle<T>(string procedureName, List<SqlParameter> parameters) where T : class
        {
            if (string.IsNullOrEmpty(_connectionString)) { throw new ArgumentNullException("_connectionString"); }
            return ReadSingle<T>(_connectionString, procedureName, parameters);
        }

        public T ReadSingle<T>(string connectionString, string procedureName) where T : class
        {
            return ReadSingle<T>(connectionString, procedureName, (List<SqlParameter>)null);
        }

        public T ReadSingle<T>(string connectionString, string procedureName, Dictionary<string, object> parameters) where T : class
        {
            var sqlParameters = new List<SqlParameter>();
            foreach (var param in parameters)
            {
                sqlParameters.Add(new SqlParameter(param.Key, param.Value));
            }
            return ReadSingle<T>(connectionString, procedureName, sqlParameters);
        }

        public T ReadSingle<T>(string connectionString, string procedureName, List<SqlParameter> parameters) where T : class
        {
            var result = ExecuteReader(connectionString, procedureName, parameters);
            var item = result.FirstOrDefault();
            if (item == null) return default(T);
            return DynamicConverter.Convert<T>(item);
        }

        public List<T> ReadList<T>(string procedureName) where T : class
        {
            if (string.IsNullOrEmpty(_connectionString)) { throw new ArgumentNullException("_connectionString"); }
            return ReadList<T>(_connectionString, procedureName, (List<SqlParameter>)null);
        }

        public List<T> ReadList<T>(string procedureName, Dictionary<string, object> parameters) where T : class
        {
            if (string.IsNullOrEmpty(_connectionString)) { throw new ArgumentNullException("_connectionString"); }
            return ReadList<T>(_connectionString, procedureName, parameters);
        }

        public List<T> ReadList<T>(string procedureName, List<SqlParameter> parameters) where T : class
        {
            if (string.IsNullOrEmpty(_connectionString)) { throw new ArgumentNullException("_connectionString"); }
            return ReadList<T>(_connectionString, procedureName, parameters);
        }

        public List<T> ReadList<T>(string connectionString, string procedureName) where T : class
        {
            return ReadList<T>(connectionString, procedureName, (List<SqlParameter>)null);
        }

        public List<T> ReadList<T>(string connectionString, string procedureName, Dictionary<string, object> parameters) where T : class
        {
            var sqlParameters = new List<SqlParameter>();
            foreach(var item in parameters)
            {
                sqlParameters.Add(new SqlParameter(item.Key, item.Value));
            }
            return ReadList<T>(connectionString, procedureName, sqlParameters);
        }

        public List<T> ReadList<T>(string connectionString, string procedureName, List<SqlParameter> parameters) where T : class
        {
            var items = ExecuteReader(connectionString, procedureName, parameters);
            var result = new List<T>();
            foreach (var item in items)
            {
                result.Add(DynamicConverter.Convert<T>(item));
            }
            return result;
        }

        public void ExecuteNonQuery(string procedureName, Dictionary<string, object> parameters)
        {
            if (string.IsNullOrEmpty(_connectionString)) throw new ArgumentNullException("_connectionString");
            ExecuteNonQuery(_connectionString, procedureName, parameters);
        }

        public void ExecuteNonQuery(string procedureName, List<SqlParameter> parameters)
        {
            if (string.IsNullOrEmpty(_connectionString)) throw new ArgumentNullException("_connectionString");
            ExecuteNonQuery(_connectionString, procedureName, parameters);
        }

        public void ExecuteNonQuery(string connectionString, string procedureName, Dictionary<string, object> parameters)
        {
            var sqlParameters = new List<SqlParameter>();
            if (parameters != null)
            {
                foreach (var param in parameters)
                {
                    sqlParameters.Add(new SqlParameter(param.Key, param.Value));
                }
            }
            ExecuteNonQuery(connectionString, procedureName, sqlParameters);
        }

        public void ExecuteNonQuery(string connectionString, string procedureName, List<SqlParameter> parameters)
        {
            using (var con = new SqlConnection(connectionString))
            using (var cmd = new SqlCommand())
            {
                cmd.Connection = con;
                cmd.CommandText = procedureName;
                cmd.CommandType = System.Data.CommandType.StoredProcedure;
                cmd.Parameters.Clear();
                if (_standardCommandParameters != null)
                {
                    foreach (var param in _standardCommandParameters)
                    {
                        cmd.Parameters.Add(param);
                    }
                }
                if (parameters != null)
                {
                    foreach (var param in parameters)
                    {
                        cmd.Parameters.Add(param);
                    }
                }
                try
                {
                    con.Open();
                    cmd.ExecuteNonQuery();
                    if (!string.IsNullOrEmpty(cmd.Parameters["@ErrNum"].Value.ToString()) && cmd.Parameters["@ErrNum"].Value.ToString() != "0")
                    {
                        var errNum = (int)cmd.Parameters["@ErrNum"].Value;
                        var errText = cmd.Parameters["@ErrText"].Value.ToString();
                        throw new SqlCmdException()
                        {
                            ErrNum = errNum,
                            ErrText = errText
                        };
                    }
                }
                catch
                {
                    throw;
                }
            }
        }

        public void ExecuteNonQuery<T>(string connectionString, string procedureName, T item) where T : class
        {
            throw new NotImplementedException();
        }

        #endregion Public methods

        #region Private methods

        private List<dynamic> ExecuteReader(string connectionString, string procedureName, List<SqlParameter> parameters)
        {
            using (var con = new SqlConnection(connectionString))
            using (var cmd = new SqlCommand())
            {
                cmd.Connection = con;
                cmd.CommandText = procedureName;
                cmd.CommandType = System.Data.CommandType.StoredProcedure;
                cmd.Parameters.Clear();
                if (_standardCommandParameters != null)
                {
                    foreach (var param in _standardCommandParameters)
                    {
                        //if (cmd.Parameters.Contains(param.ParameterName)) { throw new Exception(param.ParameterName); }
                        var newParam = new SqlParameter(param.ParameterName, param.SqlDbType, param.Size)
                        {
                            Direction = param.Direction,
                            Value = param.Value
                        };
                        cmd.Parameters.Add(newParam);
                    }
                }
                if (parameters != null)
                {
                    foreach (var param in parameters)
                    {
                        if (cmd.Parameters.Contains(param.ParameterName)) { throw new Exception(param.ParameterName); }
                        cmd.Parameters.Add(param);
                    }
                }
                try
                {
                    con.Open();
                    using (var reader = cmd.ExecuteReader())
                    {
                        var result = new List<dynamic>();

                        while (reader.Read())
                        {
                            var newItem = new ExpandoObject() as IDictionary<string, object>;
                            for (int i = 0; i < reader.FieldCount; i++)
                            {
                                var type = reader[i].GetType();
                                newItem.Add(reader.GetName(i), type != typeof(DBNull) ? reader[i] : null);
                            }
                            result.Add(newItem);
                        }
                        //if (result.Count == 1)
                        //{
                        //    return result.First();
                        //}
                        return result;
                    }
                }
                catch
                {
                    throw;
                }
            }
        }

        #endregion Private methods
    }
}

File: ToWer.DbWrapper.VB/MsSql/MsSqlWrapper.vb

Imports System.Data.SqlClient
Imports System.Dynamic
Imports ToWer.DbWrapper.VB.Converter

Namespace MsSql


    Public Class MsSqlWrapper
        Implements IDbWrapper

        Public Function ReadSingle(Of T As Class)(ByVal connectionString As String, ByVal procedureName As String) As T Implements IDbWrapper.ReadSingle
            Return ReadSingle(Of T)(connectionString, procedureName, Nothing)
        End Function

        Public Function ReadSingle(Of T As Class)(ByVal connectionString As String, ByVal procedureName As String, ByVal parameters As Dictionary(Of String, Object)) As T Implements IDbWrapper.ReadSingle
            Dim result = ExecuteReader(connectionString, procedureName, parameters)
            Dim item = result.FirstOrDefault()
            If item Is Nothing Then Return Nothing
            Return DynamicConverter.Convert(Of T)(item)
        End Function

        Public Function ReadList(Of T As Class)(ByVal connectionString As String, ByVal procedureName As String) As List(Of T) Implements IDbWrapper.ReadList
            Return ReadList(Of T)(connectionString, procedureName, Nothing)
        End Function

        Public Function ReadList(Of T As Class)(ByVal connectionString As String, ByVal procedureName As String, ByVal parameters As Dictionary(Of String, Object)) As List(Of T) Implements IDbWrapper.ReadList
            Dim items = ExecuteReader(connectionString, procedureName, parameters)
            Dim result = New List(Of T)()
            For Each item In items
                result.Add(DynamicConverter.Convert(Of T)(item))
            Next

            Return result
        End Function

        Public Sub ExecuteNonQuery(ByVal connectionString As String, ByVal procedureName As String, ByVal parameters As Dictionary(Of String, Object)) Implements IDbWrapper.ExecuteNonQuery
            Using con = New SqlConnection(connectionString)
                Using cmd = New SqlCommand()
                    cmd.Connection = con
                    cmd.CommandText = procedureName
                    cmd.CommandType = System.Data.CommandType.StoredProcedure
                    If parameters IsNot Nothing Then
                        For Each param In parameters
                            cmd.Parameters.AddWithValue("@" & param.Key, param.Value)
                        Next
                    End If

                    Try
                        con.Open()
                        cmd.ExecuteNonQuery()
                    Catch
                        Throw
                    End Try
                End Using
            End Using
        End Sub

        Public Sub ExecuteNonQuery(Of T As Class)(ByVal connectionString As String, ByVal procedureName As String, ByVal item As T) Implements IDbWrapper.ExecuteNonQuery
            Throw New NotImplementedException()
        End Sub

        Private Function ExecuteReader(ByVal connectionString As String, ByVal procedureName As String, ByVal parameters As Dictionary(Of String, Object)) As List(Of Object)
            Using con = New SqlConnection(connectionString)
                Using cmd = New SqlCommand()
                    cmd.Connection = con
                    cmd.CommandText = procedureName
                    cmd.CommandType = System.Data.CommandType.StoredProcedure
                    If parameters IsNot Nothing Then
                        For Each param In parameters
                            cmd.Parameters.AddWithValue("@" & param.Key, param.Value)
                        Next
                    End If

                    Try
                        con.Open()
                        Using reader = cmd.ExecuteReader()
                            Dim result = New List(Of Object)()
                            While reader.Read()
                                Dim newItem = TryCast(New ExpandoObject(), IDictionary(Of String, Object))
                                For i As Integer = 0 To reader.FieldCount - 1
                                    Dim type = reader(i).[GetType]()
                                    newItem.Add(reader.GetName(i), If(type <> GetType(DBNull), reader(i), Nothing))
                                Next

                                result.Add(newItem)
                            End While

                            Return result
                        End Using
                    Catch
                        Throw
                    End Try
                End Using
            End Using
        End Function
    End Class

End Namespace